pkg/util/cache/intern.go

```diff
@@ -0,0 +1,359 @@
+		fileBacking:    backing,
+		maxStringCount: maxStringCount,
+		refcount:       1,
+	}
+	log.Debugf("Created new String interner %p with mmap hash %p with max size %d", i, backing, maxStringCount)
+	return i
+}
+
+// loadOrStore always returns the string from the cache, adding it into the
+// cache if needed.
+// If we need to store a new entry and the cache is at its maximum capacity,
+// it is reset.  The origin identifies the container (and thusly, the quota)
+// originating the string.
+func (i *stringInterner) loadOrStore(key []byte) string {
+	if len(key) == 0 {
+		// Empty key
+		return ""
+	}
+	if i.maxStringCount == 0 {
+		// Dead interner.  release() has already been called, or it was broken on
+		// construction.
+		return ""
+	}
+
+	result := i.cache.lookupOrInsert(key, func(key []byte) string {
+		if i.fileBacking != nil {
+			s, possible := i.fileBacking.lookupOrInsert(key)
+			if len(s) == 0 {
+				if !possible {
+					// String is too large to allocate in the backing, just do it
+					// on the heap.  Let GC take care of it normally.  The string cache
+					// already de-duplicates for us.
+					return string(key)
+				}
+				// Return the empty string here, and "loadOrStore's" caller will
+				// resize the interner in response.
+				return ""
+			}
+			return s
+		}
+		return string(key)
+	})
+
+	return result
+}
+
+func (i *stringInterner) used() int64 {
+	used, _ := i.fileBacking.sizes()
+	return used
+}
+
+// Retain some references
+func (i *stringInterner) Retain(n int32) {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	i.refcount += n
+}
+
+// Release some references
+func (i *stringInterner) Release(n int32) {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	if i.refcount < 1 {
+		log.Warnf("Dead stringInterner being released!  refcount=%d", i.refcount)
+		return
+	}
+	if i.refcount > 0 && i.refcount-n < 1 {
+		log.Debugf("Finalizing backing, refcount=%d, n=%d", i.refcount, n)
+		i.fileBacking.finalize()
+		i.cache = newLruStringCache(0, false)
+		i.maxStringCount = 0
+	}
+	i.refcount -= n
+}
+
+func (i *stringInterner) retain() {
+	i.refcountLock.Lock()
+	defer i.refcountLock.Unlock()
+	if i.refcount < 1 {
+		log.Error("Dead interner being re-retained!")
+	}
+	i.refcount++
+}
+
+// KeyedInterner has an origin-keyed set of interners.
+type KeyedInterner struct {
+	interners      *lru.Cache
+	maxQuota       int
+	closeOnRelease bool
+	tmpPath        string
+	lastReport     time.Time
+	minFileSize    int64
+	maxPerInterner int
+	lock           sync.Mutex
+}
+
+// NewKeyedStringInterner creates a Keyed String Interner with a max per-origin quota of maxQuota
+func NewKeyedStringInterner(cfg cconfig.Component) *KeyedInterner {
+	stringInternerCacheSize := cfg.GetInt("dogstatsd_string_interner_size")
+	enableMMap := cfg.GetBool("dogstatsd_string_interner_mmap_enable")
+
+	if enableMMap {
+		closeOnRelease := !cfg.GetBool("dogstatsd_string_interner_mmap_preserve")
```

Comment:
We're missing the documentation for all these no ?

Reply:
Yeah, where should the docs go?  I didn't see much where I defined them:
https://github.com/DataDog/datadog-agent/blob/lally/exp-mem-metrefs/pkg/config/config.go#L540-L545

Reply:
They've since been added.